src/toxcore/dht/server.rs

```diff
@@ -261,114 +255,141 @@ impl Server {
-                    id: payload.id,
-                };
-                client.send_nodes_response(resp_payload)
-            } else {
-                error!("get_closest() return nothing");
-                Box::new( future::err(
-                    Error::new(ErrorKind::Other,
-                        "get_closest() return nothing"
-                )))
-            }
-        }
-        else {
+    fn handle_nodes_req(&self, packet: NodesRequest, addr: SocketAddr) -> IoFuture<()> {
+        let state = self.state.read();
+        let client = self.create_client(&addr, packet.pk, &state.peers_cache);
+
+        let payload = packet.get_payload(&self.sk);
+        if payload.is_err() {
             error!("get_payload() fail upon NodesRequest");
-            Box::new( future::err(
+            return Box::new( future::err(
                 Error::new(ErrorKind::Other,
                     "get_payload() fail upon NodesRequest"
             )))
         }
+        let payload = payload.unwrap();
+
+        let close_nodes = state.kbucket.get_closest(&self.pk);
+        if !close_nodes.is_empty() {
+            let resp_payload = NodesResponsePayload {
+                nodes: close_nodes,
+                id: payload.id,
+            };
+            client.send_nodes_response(resp_payload)
+        } else {
+            error!("get_closest() return nothing");
+            Box::new( future::err(
+                Error::new(ErrorKind::Other,
+                    "get_closest() return nothing"
+            )))
+        }
     }
     /**
     handle received NodesResponse from peer.
     */
-    fn handle_nodes_resp(&self, client: Client, request: NodesResponse) -> IoFuture<()> {
-        if let Ok(payload) = request.get_payload(&self.sk) {
-            if payload.id == 0 {
-                return Box::new( future::err(
-                    Error::new(ErrorKind::Other,
-                        "NodesResponse.ping_id == 0"
-                )))
-            }
-            if client.ping_id == payload.id {
-                let mut state = self.state.write();
-                for node in &payload.nodes {
-                    state.kbucket.try_add(node);
-                }
-                Box::new( future::ok(()) )
-            }
-            else {
-                Box::new( future::err(
-                    Error::new(ErrorKind::Other, "NodesResponse.ping_id does not match")
-                ))
-            }
-        }
-        else {
+    fn handle_nodes_resp(&self, packet: NodesResponse) -> IoFuture<()> {
+        let mut state = self.state.write();
+        let payload = packet.get_payload(&self.sk);
+        if payload.is_err() {
             error!("get_payload() fail upon NodesResponse");
-            Box::new( future::err(
+            return Box::new( future::err(
                 Error::new(ErrorKind::Other,
                     "get_payload() fail upon NodesResponse"
             )))
         }
+        let payload = payload.unwrap();
```

Comment:
You can use `match` to avoid unwrap here, something like:
```rust
let payload = payload match {
    Ok(payload) => payload,
    Err(_) => { ... }
}
```

Reply:
Actually he took it from my code. Well... @NamsooCho you may use this trick, but `Err` first. And it is up to you.

Reply:
Using unwrap is a bad style when it's avoidable :)